File: Data/C++/subset generator power set.cpp
#include <iostream>
#include <vector>
using namespace std;

void generateSubsets(int i, vector<int>& nums, vector<int>& current, vector<vector<int>>& result) {
    if (i == nums.size()) {
        result.push_back(current);  // Reached end, store current subset
        return;
    }

    // Include nums[i]
    current.push_back(nums[i]);
    generateSubsets(i + 1, nums, current, result);

    // Exclude nums[i]
    current.pop_back();
    generateSubsets(i + 1, nums, current, result);
}

// ✅ Use-case
int main() {
    vector<int> nums = {1, 2, 3};
    vector<vector<int>> result, current;

    generateSubsets(0, nums, current, result);

    for (auto& subset : result) {
        cout << "[ ";
        for (int x : subset) cout << x << " ";
        cout << "]\n";
    }
}

File: Data/Python/subset generator power set.py
def generate_subsets(nums):
    """
    Generate all subsets (power set):
    - Either include or exclude current element.
    - Backtracking used to explore all choices.
    - Time Complexity: O(2^n)
    """
    result = []

    def backtrack(index, path):
        if index == len(nums):
            result.append(path[:])  # Add a copy of current subset
            return
        # Exclude current
        backtrack(index + 1, path)
        # Include current
        path.append(nums[index])
        backtrack(index + 1, path)
        path.pop()  # Backtrack

    backtrack(0, [])
    return result

# ✅ Example
print("Subsets:", generate_subsets([1, 2, 3]))
# Output: All subsets of [1, 2, 3]
